src/cluster.c

```diff
@@ -5147,7 +5147,11 @@ void clusterSetMaster(clusterNode *n) {
     serverAssert(n != myself);
     serverAssert(myself->numslots == 0);
 
-    removeAllShardChannelSubscription();
+    /* If the node is still in the same shard, don't need to unsubscribe the clients. */
+    if (memcmp(myself->shard_id, n->shard_id, CLUSTER_NAMELEN) != 0) {
+            removeAllShardChannelSubscription();
+    }
```

Comment:
If you want this check, then why not do it in `clusterRemoveNodeFromShard`? (called from this function via updateShardId a few lines below)

(If you don't want the check, then I agree clusterSetMaster is the right place.)

Reply:
As I explained earlier, the problem with putting it under `clusterRemoveNodeFromShard` is the `slaveof` value has already changed to the new master and the subscription removal logic depends on the old master info to cleanup the active slots.

Reply:
Ah, ok